src/resolve/source.rs

```diff
@@ -66,52 +97,70 @@ impl Source {
     }
 
-    pub fn into_include(self, remote: &Remote) -> io::Result<Include> {
+    pub fn into_include(
+        self, remote: &Remote, range: Range<usize>, diagnostics: &mut Diagnostics<'_>,
+    ) -> Result<Include> {
         let Source { url, group } = self;
         match group {
             SourceGroup::Implementation => {
                 if let Some(domain) = url.domain() {
                     if let Ok(command) = Command::from_str(domain) {
                         Ok(Include::Command(command))
                     } else {
-                        Err(io::Error::new(
-                            io::ErrorKind::NotFound,
-                            format!("No heradoc implementation found for domain {:?}", domain),
-                        ))
+                        diagnostics
+                            .error(format!(
+                                "no heradoc implementation found for domain {:?}",
+                                domain
+                            ))
+                            .with_error_section(&range, "defined here")
+                            .emit();
+                        Err(Error::Diagnostic)
                     }
                 } else {
-                    Err(io::Error::new(
-                        io::ErrorKind::NotFound,
-                        "No heradoc implementation domain found",
-                    ))
+                    diagnostics
+                        .error("no heradoc implementation domain found")
+                        .with_error_section(&range, "defined here")
+                        .emit();
+                    Err(Error::Diagnostic)
                 }
             },
             SourceGroup::LocalRelative(path) => {
                 let parent = path.parent().unwrap().to_owned();
-                to_include(path, Context::LocalRelative(parent))
+                to_include(path, Context::LocalRelative(parent), range, diagnostics)
             },
             SourceGroup::LocalAbsolute(path) => {
                 let parent = path.parent().unwrap().to_owned();
-                to_include(path, Context::LocalAbsolute(parent))
+                to_include(path, Context::LocalAbsolute(parent), range, diagnostics)
             },
             SourceGroup::Remote => {
                 let downloaded = match remote.http(&url) {
                     Ok(downloaded) => downloaded,
-                    Err(RemoteError::Io(io)) => return Err(io),
-                    // TODO: proper error handling with failure
-                    Err(RemoteError::Request(_req)) => {
-                        return Err(io::ErrorKind::ConnectionAborted.into());
+                    Err(RemoteError::Io(err)) => {
+                        diagnostics
+                            .error("error writing downloaded content to cache")
+                            .with_error_section(&range, "trying to download this")
+                            .error(format!("cause: {}", err))
```

Comment:
Also include the path the cached file should be written to, this would need to be propagated to the `Error` from `Remote`. It can not be resolve before, as the content type depends on the server response and possible cache presence.

This can be opened as a separate issue if you want.